File: backend/app/api/sse.py
"""SSE 适配层：把进程内 Lifecycle 事件翻译成前端可读的 SSE 事件。

ChatService 与 AgentRuntime 只发布领域生命周期事件，不直接操作 HTTP；
本模块是“事件 → 传输格式（event: xxx + data: {...}）”的唯一映射点，
未来同一份事件还可以驱动日志、指标、评估等其它 Adapter。
Tool 执行进度正式映射为 tool.started / tool.completed；ToolRuntime
不直接发送 SSE。
"""

import json
from collections.abc import AsyncIterator

from app.agent.lifecycle.events import (
    LifecycleEvent,
    ReasoningStarted,
    ResponseDelta,
    ToolCompleted,
    ToolStarted,
    TurnCancelled,
    TurnCommitted,
    TurnFailed,
    TurnStarted,
)


def map_lifecycle_event(event: LifecycleEvent) -> tuple[str, dict[str, object]] | None:
    """把生命周期事件映射为（SSE 事件名, 载荷）；返回 None 表示该事件不推送。"""
    if isinstance(event, TurnStarted):
        return "run.started", {
            "turn_id": str(event.turn_id),
            "run_id": str(event.run_id),
            "thread_id": str(event.thread_id),
        }
    if isinstance(event, ReasoningStarted):
        return "reasoning.started", {
            "turn_id": str(event.turn_id),
            "run_id": str(event.run_id),
            "step_index": event.step_index,
        }
    if isinstance(event, ResponseDelta):
        # 流式正文增量；step_index 供前端跨推理步切换缓冲（兜底用）
        return "response.delta", {
            "content": event.delta,
            "step_index": event.step_index,
        }
    if isinstance(event, ToolStarted):
        return "tool.started", {
            "tool": event.tool,
            "call_id": str(event.call_id),
        }
    if isinstance(event, ToolCompleted):
        return "tool.completed", {
            "tool": event.tool,
            "call_id": str(event.call_id),
            "status": event.status,
            "duration_ms": event.duration_ms,
        }
    if isinstance(event, TurnCommitted):
        return "run.completed", {
            "turn_id": str(event.turn_id),
            "run_id": str(event.run_id),
            "message_id": str(event.assistant_message_id),
        }
    if isinstance(event, TurnFailed):
        return "run.failed", {"turn_id": str(event.turn_id), "error": event.error}
    if isinstance(event, TurnCancelled):
        return "run.cancelled", {"turn_id": str(event.turn_id)}
    # 未知事件类型：跳过不推送（向前兼容新增事件）。
    return None


def format_sse(event_name: str, payload: dict[str, object]) -> str:
    """编码一帧 SSE（Server-Sent Events）：event 行 + data 行 + 空行结尾。"""
    return f"event: {event_name}\ndata: {json.dumps(payload, ensure_ascii=False)}\n\n"


async def iter_sse(chunks: AsyncIterator[str]) -> AsyncIterator[str]:
    """把异步字符串流原样透传为 SSE 流（当前为直通实现）。"""
    async for chunk in chunks:
        yield chunk

File: frontend/src/lib/sse.ts
/**
 * SSE 客户端：POST /api/v1/chat/stream 的流式解析。
 *
 * EventSource 无法携带 Authorization 头，这里用 fetch + ReadableStream
 * 手工解析 `event: <名>\\ndata: <json>\\n\\n` 帧（与 backend/app/api/sse.py
 * 的格式对应）。response.delta 是流式正文增量：随模型生成逐片段推送，
 * step_index 标识产生增量的推理步，跨步时应清空缓冲重新累积。
 */

import { ApiError, UnauthorizedError } from "@/lib/api";
import { clearToken, loadToken } from "@/lib/token";

export interface ToolTrace {
  callId: string;
  tool: string;
  status: string | null;
  durationMs: number | null;
  done: boolean;
}

export type StreamEvent =
  | { type: "run.started"; turnId: string; threadId: string }
  | { type: "reasoning.started" }
  | { type: "tool.started"; trace: ToolTrace }
  | { type: "tool.completed"; trace: ToolTrace }
  | { type: "response.delta"; content: string; stepIndex: number }
  | { type: "run.completed" }
  | { type: "run.failed"; error: string }
  | { type: "run.cancelled" };

interface WireFrame {
  turn_id?: string;
  thread_id?: string;
  run_id?: string;
  tool?: string;
  call_id?: string;
  status?: string;
  duration_ms?: number;
  content?: string;
  step_index?: number;
  message_id?: string;
  error?: string;
}

function translate(event: string, data: WireFrame): StreamEvent | null {
  switch (event) {
    case "run.started":
      return data.thread_id
        ? { type: "run.started", turnId: data.turn_id ?? "", threadId: data.thread_id }
        : null;
    case "reasoning.started":
      return { type: "reasoning.started" };
    case "tool.started":
      return data.tool && data.call_id
        ? {
            type: "tool.started",
            trace: {
              callId: data.call_id,
              tool: data.tool,
              status: null,
              durationMs: null,
              done: false,
            },
          }
        : null;
    case "tool.completed":
      return data.tool && data.call_id
        ? {
            type: "tool.completed",
            trace: {
              callId: data.call_id,
              tool: data.tool,
              status: data.status ?? null,
              durationMs: data.duration_ms ?? null,
              done: true,
            },
          }
        : null;
    case "response.delta":
      return {
        type: "response.delta",
        content: data.content ?? "",
        stepIndex: data.step_index ?? 0,
      };
    case "run.completed":
      return { type: "run.completed" };
    case "run.failed":
      return { type: "run.failed", error: data.error ?? "执行失败" };
    case "run.cancelled":
      return { type: "run.cancelled" };
    default:
      return null;
  }
}

function parseChunk(buffer: string): { events: Array<{ event: string; data: WireFrame }>; rest: string } {
  const events: Array<{ event: string; data: WireFrame }> = [];
  let rest = buffer;
  let separator = rest.indexOf("\n\n");
  while (separator !== -1) {
    const frame = rest.slice(0, separator);
    rest = rest.slice(separator + 2);
    let event = "message";
    const dataLines: string[] = [];
    for (const line of frame.split("\n")) {
      if (line.startsWith("event:")) event = line.slice(6).trim();
      else if (line.startsWith("data:")) dataLines.push(line.slice(5).trimStart());
    }
    if (dataLines.length > 0) {
      try {
        events.push({ event, data: JSON.parse(dataLines.join("\n")) as WireFrame });
      } catch {
        // 单帧解析失败只跳过该帧，不中断整个流。
      }
    }
    separator = rest.indexOf("\n\n");
  }
  return { events, rest };
}

/** 发送消息并逐事件回调。resolve 于流正常结束；HTTP/鉴权错误直接抛出。 */
export async function streamChat(
  message: string,
  threadId: string | null,
  onEvent: (event: StreamEvent) => void,
): Promise<void> {
  const token = loadToken();
  const response = await fetch("/api/v1/chat/stream", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "text/event-stream",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify({ thread_id: threadId, message }),
  });

  if (response.status === 401) {
    clearToken();
    throw new UnauthorizedError();
  }
  if (!response.ok || !response.body) {
    let message = `连接教练失败（${response.status}）`;
    try {
      const body = (await response.json()) as { detail?: unknown };
      if (typeof body?.detail === "string") message = body.detail;
    } catch {
      // 非 JSON 错误体时保留状态码信息。
    }
    throw new ApiError(response.status, message);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    const { events, rest } = parseChunk(buffer);
    buffer = rest;
    for (const frame of events) {
      const translated = translate(frame.event, frame.data);
      if (translated) onEvent(translated);
    }
  }
}
